File: app/src/main/java/eu/szwiec/replayview/replay/ReplayViewModel.java
package eu.szwiec.replayview.replay;

import android.app.Application;
import android.arch.lifecycle.AndroidViewModel;
import android.arch.lifecycle.MutableLiveData;
import android.support.annotation.NonNull;

import java.util.Arrays;
import java.util.Collections;
import java.util.List;

import rx.Subscription;
import rx.android.schedulers.AndroidSchedulers;
import rx.schedulers.Schedulers;
import timber.log.Timber;

public class ReplayViewModel extends AndroidViewModel {
    private MutableLiveData<Boolean> mIsProcessingLiveData = new MutableLiveData<>();
    private MutableLiveData<List<ReplayEvent>> mEventsLiveData = new MutableLiveData<>();

    private MutableLiveData<Boolean> mIsPlayingLiveData = new MutableLiveData<>();
    private MutableLiveData<Integer> mProgressLiveData = new MutableLiveData<>();

    public MutableLiveData<String> mPlayingTimeLiveData = new MutableLiveData<>();
    public MutableLiveData<String> mTotalTimeLiveData = new MutableLiveData<>();

    public final List<String> availableDataTypes = Collections.unmodifiableList(Arrays.asList(ImportDataManager.TYPE_WIFI, ImportDataManager.TYPE_GPS, ImportDataManager.TYPE_BLUETOOTH, ImportDataManager.TYPE_SENSOR));

    private CharSequence[] mPickedDataTypes;
    private ImportDataManager mImportDataManager;
    private Subscription mSubscription;

    public ReplayViewModel(@NonNull Application application) {
        super(application);

        mIsPlayingLiveData.setValue(false);
        mImportDataManager = new ImportDataManager(application);
    }

    public void toggle() {
        if (mIsPlayingLiveData.getValue() == true) {
            pause();
        } else {
            play();
        }
    }

    public MutableLiveData<Boolean> getIsPlayingLiveData() {
        return mIsPlayingLiveData;
    }

    public MutableLiveData<Boolean> getIsProcessingLiveData() {
        return mIsProcessingLiveData;
    }

    public MutableLiveData<List<ReplayEvent>> getEventsLiveData() {
        return mEventsLiveData;
    }

    public MutableLiveData<Integer> getProgressLiveData() {
        return mProgressLiveData;
    }

    public void zipPicked(String path) {
        mIsProcessingLiveData.setValue(true);
        importData(path, mPickedDataTypes);
    }

    private void importData(String path, CharSequence[] pickedDataTypes) {

        mSubscription = mImportDataManager.getDataObservable(path, pickedDataTypes)
                .subscribeOn(Schedulers.newThread())
                .observeOn(AndroidSchedulers.mainThread())
                .subscribe(
                        venues -> onSuccess(venues),
                        throwable -> onError(throwable)
                );
    }

    public void typesPicked(CharSequence[] pickedTypes) {
        mPickedDataTypes = pickedTypes;
    }

    private void onSuccess(List<ReplayEvent> replayEvents) {
        mEventsLiveData.setValue(replayEvents);
        mIsProcessingLiveData.setValue(false);
    }

    private void onError(Throwable throwable) {
        Timber.e(throwable);
        mIsProcessingLiveData.setValue(false);
    }

    private Thread mPlayingThread;

    private void initPlayback() {
        setProgress(0);

        mPlayingThread = new Thread(() -> {
            while (true) {
                if (mIsPlayingLiveData.getValue()) {
                    Integer progress = mProgressLiveData.getValue();
                    if(progress == 100) {
                        stop();
                    } else {
                        postProgress(progress+1);
                        try {
                            Thread.sleep(100);
                        } catch (InterruptedException e) {
                            e.printStackTrace();
                        }
                    }
                }
            }
        });
    }

    private void play() {
        if (mPlayingThread == null) {
            initPlayback();
        }

        mIsPlayingLiveData.setValue(true);
        if (mPlayingThread.getState() == Thread.State.NEW) {
            mPlayingThread.start();
        }
    }

    private void pause() {
        mIsPlayingLiveData.setValue(false);
    }

    private void stop() {
        mIsPlayingLiveData.setValue(false);
        setProgress(0);
    }

    public String getPlayingTime(int progress) {
        return progress + "";
    }

    public String getTotalTime() {
        return "100";
    }

    public void setProgress(int progress) {
        mProgressLiveData.setValue(progress);
        mPlayingTimeLiveData.setValue(getPlayingTime(progress));
    }
    public void postProgress(int progress) {
        mProgressLiveData.postValue(progress);
        mPlayingTimeLiveData.postValue(getPlayingTime(progress));
    }

}

File: app/src/main/java/eu/szwiec/replayview/replay/ReplayViewModel.kt
package eu.szwiec.replayview.replay

import android.arch.lifecycle.LiveData
import android.arch.lifecycle.Transformations
import android.arch.lifecycle.ViewModel
import android.support.annotation.VisibleForTesting
import eu.szwiec.replayview.utils.NonNullLiveData
import kotlinx.coroutines.experimental.android.UI
import kotlinx.coroutines.experimental.async
import org.apache.commons.lang3.time.DurationFormatUtils
import org.jetbrains.anko.AnkoLogger
import org.jetbrains.anko.coroutines.experimental.bg
import org.jetbrains.anko.info
import java.util.concurrent.TimeUnit

class ReplayViewModel(filesProvider: ReplayFilesProvider) : ViewModel(), AnkoLogger {

    private val playingThread: Thread
    private var pickedDataTypes = emptyList<ReplayType>()

    val stateLD = NonNullLiveData(State.NOT_READY)
    val progressLD = NonNullLiveData(0)
    val speedLD = NonNullLiveData(SPEEDS[0])
    private val eventsLD = NonNullLiveData(emptyList<ReplayEvent>())
    val isPlayingLD = NonNullLiveData(false)

    val playingTimeLD: LiveData<String> = Transformations.map(progressLD, { progress -> formatPlayingTime(progress, eventsLD.value) })
    val totalTimeLD: LiveData<String> = Transformations.map(eventsLD, { events -> formatTotalTime(events) })
    val maxProgressLD: LiveData<Int> = Transformations.map(eventsLD, { events -> events.size - 1 })
    val isPlayingEnabledLD: LiveData<Boolean> = Transformations.map(stateLD, { state -> state == State.READY })

    val filesProvider: ReplayFilesProvider

    init {
        playingThread = initPlayingThread()
        this.filesProvider = filesProvider
    }

    fun togglePlayPause() {
        if (isPlayingLD.value) {
            pause()
        } else {
            play()
        }
    }

    fun changeSpeed() {
        val nextSpeedId: Int

        val currentSpeed = speedLD.value
        val currentSpeedId = SPEEDS.indexOf(currentSpeed)
        nextSpeedId = if (currentSpeedId == SPEEDS.size - 1) {
            0
        } else {
            currentSpeedId + 1
        }
        speedLD.value = SPEEDS[nextSpeedId]
    }


    fun onOpenButtonClick() {
        stateLD.value = State.PICKING_TYPE
    }

    fun setPickedTypes(types: Array<CharSequence>) {
        pickedDataTypes = ReplayType.getTypes(types)
    }

    fun onTypePicked() {
        stateLD.value = State.PICKING_FILE
    }

    fun onFilePicked(path: String?) {
        stateLD.value = State.PROCESSING
        importData(path, pickedDataTypes)
    }

    fun dialogDismissed() {
        if (eventsLD.value.isNotEmpty())
            stateLD.value = State.READY
        else
            stateLD.value = State.NOT_READY
    }

    private fun importData(path: String?, pickedDataTypes: List<ReplayType>) {
        async(UI) {
            val replayEvents = bg {
                val files = filesProvider.provide(path, pickedDataTypes)
                info("DUPA files = $files")

                val events = ReplayEventsParser.getEvents(files)
                events.sortBy { it.nanoTimestamp }
                info("events = $events")
                events
            }
            onPostExecute(replayEvents.await())
        }
    }

    private fun onPostExecute(events: List<ReplayEvent>) {
        eventsLD.value = events

        if (events.isNotEmpty()) {
            stateLD.value = State.READY
        } else {
            stateLD.value = State.ERROR
        }
    }

    private fun play() {
        isPlayingLD.value = true
        if (playingThread.state == Thread.State.NEW) {
            playingThread.start()
        }
    }

    private fun pause() {
        isPlayingLD.postValue(false)
    }

    private fun stop() {
        isPlayingLD.postValue(false)
        progressLD.postValue(0)
    }

    private fun initPlayingThread(): Thread {
        return Thread {
            while (true) {
                if (isPlayingLD.value) {
                    val progress = progressLD.value
                    val events = eventsLD.value
                    val event = events[progress]

                    //TODO post on the BUS

                    if (progress == maxProgressLD.value) {
                        stop()
                    } else {
                        progressLD.postValue(progress + 1)
                        val diffTimeMs = events[progress + 1].msTimestamp() - events[progress].msTimestamp()
                        Thread.sleep(diffTimeMs / speedLD.value)
                    }
                }
            }
        }
    }

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    fun formatPlayingTime(progress: Int, events: List<ReplayEvent>): String {
        if (events.isEmpty()) return ""

        val firstTimestamp = events[0].msTimestamp()
        val currentTimestamp = events[progress].msTimestamp()

        return format(currentTimestamp - firstTimestamp)
    }

    @VisibleForTesting(otherwise = VisibleForTesting.PRIVATE)
    fun formatTotalTime(events: List<ReplayEvent>): String {
        if (events.isEmpty()) return ""

        val firstTimestamp = events[0].msTimestamp()
        val lastTimestamp = events[events.size - 1].msTimestamp()

        return format(lastTimestamp - firstTimestamp)
    }

    private fun format(timestampMs: Long): String {
        return DurationFormatUtils.formatDuration(timestampMs, "mm:ss")
    }

    private fun ReplayEvent.msTimestamp(): Long {
        return TimeUnit.NANOSECONDS.toMillis(this.nanoTimestamp)
    }
}